dartagnan/src/main/java/com/dat3m/dartagnan/witness/WitnessGraph.java

```diff
@@ -2,156 +2,198 @@
-		if (operand instanceof BooleanFormula bool) {
-			return switch (value) {
-				case "false", "0" -> formulaManager.getBooleanFormulaManager().not(bool);
-				default -> bool;
-			};
-		}
-		BigInteger integerValue = switch (value) {
-			case "false" -> BigInteger.ZERO;
-			case "true" -> BigInteger.ONE;
-			default -> new BigInteger(value);
-		};
-		if (operand instanceof NumeralFormula.IntegerFormula integer) {
-			IntegerFormulaManager imgr = formulaManager.getIntegerFormulaManager();
-			return imgr.equal(integer, imgr.makeNumber(integerValue));
-		}
-		assert operand instanceof BitvectorFormula;
-		BitvectorFormula bitvector = (BitvectorFormula) operand;
-		BitvectorFormulaManager bvmgr = formulaManager.getBitvectorFormulaManager();
-		return bvmgr.equal(bitvector, bvmgr.makeBitvector(bvmgr.getLength(bitvector), integerValue));
-	}
-
-	public void write() {
-		try (FileWriter fw = new FileWriter(String.format("%s/witness.graphml", getOrCreateOutputDirectory()))) {
-			fw.write("<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n");
-			fw.write("<graphml xmlns=\"http://graphml.graphdrawing.org/xmlns\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\">\n");
-			for (GraphAttributes attr : GraphAttributes.values()) {
-				fw.write("<key attr.name=\"" + attr.toString() + "\" attr.type=\"string\" for=\"graph\" id=\"" + attr + "\"/>\n");
-			}
-			for (NodeAttributes attr : NodeAttributes.values()) {
-				fw.write("<key attr.name=\"" + attr.toString() + "\" attr.type=\"boolean\" for=\"node\" id=\"" + attr + "\"/>\n");
-			}
-			for (EdgeAttributes attr : EdgeAttributes.values()) {
-				fw.write("<key attr.name=\"" + attr.toString() + "\" attr.type=\"string\" for=\"edge\" id=\"" + attr + "\"/>\n");
-			}
-			fw.write(toXML());
-			fw.write("</graphml>\n");
-		} catch (IOException e1) {
-			e1.printStackTrace();
-		}
-	}
+    private final SortedSet<Node> nodes = new TreeSet<>();
+    // The order in which we add / traverse edges is important, thus a List
+    private final List<Edge> edges = new ArrayList<>();
+
+    public boolean isEmpty() {
+        return edges.isEmpty();
+    }
+
+    public void addNode(String id) {
+        nodes.add(new Node(id));
+    }
+
+    public boolean hasNode(String id) {
+        return nodes.stream().anyMatch(n -> n.getId().equals(id));
+    }
+
+    public Node getNode(String id) {
+        return nodes.stream().filter(n -> n.getId().equals(id)).findFirst().get();
+    }
+
+    public Node getEntryNode() {
+        assert(nodes.stream().anyMatch(n -> n.hasAttributed(ENTRY.toString())));
+        return nodes.stream().filter(n -> n.hasAttributed(ENTRY.toString())).findFirst().get();
+    }
+
+    public Node getViolationNode() {
+        assert(nodes.stream().anyMatch(n -> n.hasAttributed(VIOLATION.toString())));
+        return nodes.stream().filter(n -> n.hasAttributed(VIOLATION.toString())).findAny().get();
+    }
```

Comment:
You haven't touched this code in the PR, but I think this should still be fixed.
Instead of asserting the existence, you can just use `.findAny/First().orElseThrow(...)`.